File: app/src/main/java/de/markusfisch/android/shadereditor/adapter/ErrorAdapter.java
package de.markusfisch.android.shadereditor.adapter;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.ListAdapter;
import androidx.recyclerview.widget.RecyclerView;

import java.util.Locale;

import de.markusfisch.android.shadereditor.R;
import de.markusfisch.android.shadereditor.opengl.ShaderError;

public class ErrorAdapter extends ListAdapter<ShaderError, ErrorAdapter.ViewHolder> {
	@FunctionalInterface
	public interface OnItemClickListener {
		void onItemClick(int lineNumber);
	}

	@NonNull
	private final OnItemClickListener listener;

	public ErrorAdapter(@NonNull OnItemClickListener listener) {
		super(DIFF_CALLBACK);
		this.listener = listener;
	}

	@NonNull
	@Override
	public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
		return new ViewHolder(LayoutInflater.from(parent.getContext())
				.inflate(R.layout.error_item, parent, false));
	}

	@Override
	public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
		ShaderError error = getItem(position);
		if (error != null) {
			holder.update(error, listener);
		}
	}

	public static class ViewHolder extends RecyclerView.ViewHolder {
		TextView errorLine;
		TextView errorMessage;

		public ViewHolder(@NonNull View itemView) {
			super(itemView);
			errorLine = itemView.findViewById(R.id.error_line);
			errorMessage = itemView.findViewById(R.id.error_message);
		}

		public void update(@NonNull ShaderError error, @NonNull ErrorAdapter.OnItemClickListener listener) {
			if (error.hasLine()) {
				errorLine.setText(String.format(Locale.getDefault(), "%d: ", error.getLine()));
			} else {
				errorLine.setVisibility(View.GONE);
			}

			errorMessage.setText(error.getMessage());

			itemView.setOnClickListener(v -> listener.onItemClick(error.getLine()));
		}
	}

	private static final DiffUtil.ItemCallback<ShaderError> DIFF_CALLBACK =
			new DiffUtil.ItemCallback<ShaderError>() {
				@Override
				public boolean areItemsTheSame(@NonNull ShaderError oldItem,
						@NonNull ShaderError newItem) {
					// Implement logic to check if items are the same
					return oldItem.equals(newItem);
				}

				@Override
				public boolean areContentsTheSame(@NonNull ShaderError oldItem,
						@NonNull ShaderError newItem) {
					// Implement logic to check if item contents are the same
					return oldItem.equals(newItem);
				}
			};
}

File: app/src/main/java/de/markusfisch/android/shadereditor/adapter/ErrorAdapter.kt
package de.markusfisch.android.shadereditor.adapter

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import de.markusfisch.android.shadereditor.R
import de.markusfisch.android.shadereditor.opengl.ShaderError
import java.util.Locale

class ErrorAdapter(private val listener: OnItemClickListener) :
    ListAdapter<ShaderError, ErrorAdapter.ViewHolder>(DIFF_CALLBACK) {
    fun interface OnItemClickListener {
        fun onItemClick(position: Int)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val view = LayoutInflater.from(parent.context).inflate(R.layout.error_item, parent, false)
        return ViewHolder(view)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val error = getItem(position)
        error?.let {
            holder.update(it, listener)
        }
    }

    class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        private val errorLine: TextView = itemView.findViewById(R.id.error_line)
        private val errorMessage: TextView = itemView.findViewById(R.id.error_message)

        fun update(error: ShaderError, listener: OnItemClickListener) {
            if (error.hasLine()) {
                errorLine.text = String.format(Locale.getDefault(), "%d: ", error.line)
                errorLine.visibility = View.VISIBLE
            } else {
                errorLine.visibility = View.GONE
            }

            errorMessage.text = error.message

            itemView.setOnClickListener {
                listener.onItemClick(error.line)
            }
        }
    }

    companion object {
        private val DIFF_CALLBACK = object : DiffUtil.ItemCallback<ShaderError>() {
            override fun areItemsTheSame(oldItem: ShaderError, newItem: ShaderError): Boolean {
                // Check if the two items are the same
                return oldItem == newItem
            }

            override fun areContentsTheSame(oldItem: ShaderError, newItem: ShaderError): Boolean {
                // Check if the content of the two items are the same
                return oldItem == newItem
            }
        }
    }
}
